library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java

```diff
@@ -0,0 +1,172 @@
+		ProfileIndependentS2CMessage message = new ProfileIndependentS2CMessage(client.player, true, packet);
+		QuiltChatEvents.BEFORE_PROCESS.invoke(message);
+	}
+
+	@Inject(method = "onProfileIndependentMessage", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/ClientChatListener;m_jytgvbam(Lnet/minecraft/text/Text;Lnet/minecraft/network/message/MessageType$Parameters;)V", shift = At.Shift.AFTER))
+	public void quilt$afterInboundProfileIndependentMessage(ProfileIndependentMessageS2CPacket packet, CallbackInfo ci) {
+		ProfileIndependentS2CMessage message = new ProfileIndependentS2CMessage(client.player, true, packet);
+		QuiltChatEvents.AFTER_PROCESS.invoke(message);
+	}
+
+	@ModifyVariable(method = "m_fzlgisyq", at = @At("HEAD"), argsOnly = true)
+	public String quilt$modifyOutboundRawChatMessage(String string) {
+		// Not sure *why* this would be null but IDEA is complaining, so, safety first?
+		if (client.player == null) {
+			return string;
+		}
+
+		RawChatC2SMessage message = new RawChatC2SMessage(client.player, true, string);
+		return ((RawChatC2SMessage)QuiltChatEvents.MODIFY.invoke(message, message)).serialized();
+	}
+
+	@Inject(method = "m_fzlgisyq", at = @At(value = "HEAD"), cancellable = true)
+	public void quilt$cancelAndBeforeOutboundRawChatMessage(String string, CallbackInfo ci) {
+		if (client.player == null) {
+			return;
+		}
+
+		RawChatC2SMessage message = new RawChatC2SMessage(client.player, true, string);
+		if (QuiltChatEvents.CANCEL.invoke(message)) {
+			ci.cancel();
+		} else {
+			QuiltChatEvents.BEFORE_PROCESS.invoke(message);
+		}
+	}
+
+	@Inject(method = "m_fzlgisyq", at = @At(value = "TAIL"))
+	public void quilt$afterOutboundRawChatMessage(String string, CallbackInfo ci) {
+		if (client.player == null) {
+			return;
+		}
+
+		RawChatC2SMessage message = new RawChatC2SMessage(client.player, true, string);
+		QuiltChatEvents.AFTER_PROCESS.invoke(message);
+	}
+
+	@Redirect(method = "m_fzlgisyq", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/network/ClientPlayNetworkHandler;sendPacket(Lnet/minecraft/network/Packet;)V"))
+	public void quilt$modifyAndCancelAndBeforeAndAfterOutboundChatMessage(ClientPlayNetworkHandler instance, Packet<?> packet) {
+		if (packet instanceof ChatMessageC2SPacket chatMessageC2SPacket) {
+			ChatC2SMessage message = new ChatC2SMessage(client.player, true, chatMessageC2SPacket);
+			message = (ChatC2SMessage)QuiltChatEvents.MODIFY.invoke(message, message);
+
+			if (QuiltChatEvents.CANCEL.invoke(message) != Boolean.TRUE) {
+				QuiltChatEvents.BEFORE_PROCESS.invoke(message);
+				instance.sendPacket(message.serialized());
+				QuiltChatEvents.AFTER_PROCESS.invoke(message);
+			}
+		} else {
+			throw new IllegalArgumentException("Received non-ChatMessageC2SPacket for argument to ClientPlayNetworkHandler.sendPacket in ClientPlayNetworkHandler.m_fzlgisyq (sendChatMessage? mapping missing at time of writing)");
+		}
```

Comment:
Hm, I do wonder if the else should be removed or not

Reply:
I left the `else` in for these cases because the control flow felt more english-like to me. "if of type, do x, otherwise throw

Reply:
ah, I see